### PR TITLE
Add support of parsing OPTIMIZE TABLE statement for ClickHouse

### PR DESCRIPTION
**Pull request by git-hulk**
ClickHouse supports using `OPTIMIZE TABLE` to initialize an unscheduled merge of data parts for tables. The syntax is following:

```SQL
OPTIMIZE TABLE [db.]name [ON CLUSTER cluster] [PARTITION partition | PARTITION ID 'partition_id'] [FINAL] [DEDUPLICATE [BY expression]]
```

See: https://clickhouse.com/docs/en/sql-reference/statements/optimize

### CODE REVIEW

File: src/ast/ddl.rs [hunk 1]
```diff
@@ -1296,20 +1296,43 @@ impl fmt::Display for UserDefinedTypeCompositeAttributeDef {
...
 }
 
-/// PARTITION statement used in ALTER TABLE et al. such as in Hive SQL
+/// PARTITION statement used in ALTER TABLE et al. such as in Hive and ClickHouse SQL
 #[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "visitor", derive(Visit, VisitMut))]
-pub struct Partition {
-    pub partitions: Vec<Expr>,
+pub enum Partition {
```
Comment (git-hulk) [inline src/ast/ddl.rs:1303]: I changed the type of `Partition` from `struct` to `enum` to support more parsing states.

**Comment by git-hulk** [comment]
cc @iffyio

File: src/ast/ddl.rs [hunk 2]
```diff
@@ -1296,20 +1296,43 @@ impl fmt::Display for UserDefinedTypeCompositeAttributeDef {
     }
 }
 
-/// PARTITION statement used in ALTER TABLE et al. such as in Hive SQL
+/// PARTITION statement used in ALTER TABLE et al. such as in Hive and ClickHouse SQL
```
Comment (iffyio) [inline src/ast/ddl.rs:1299]: we could add the clickhouse link here as well?

File: src/ast/ddl.rs [hunk 3]
```diff
@@ -1296,20 +1296,43 @@ impl fmt::Display for UserDefinedTypeCompositeAttributeDef {
...
 #[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "visitor", derive(Visit, VisitMut))]
-pub struct Partition {
-    pub partitions: Vec<Expr>,
+pub enum Partition {
+    ID(Ident),
+    Expression(Expr),
```
Comment (iffyio) [inline src/ast/ddl.rs:1305]:
```suggestion
    Identifier(Ident),
    Expr(Expr),
```
I think these are more popular variant names in the repo for similar types, figured might be good to try to match them

File: src/ast/mod.rs [hunk 4]
```diff
@@ -2831,6 +2831,18 @@ pub enum Statement {
...
+    /// ```sql
+    /// OPTIMIZE TABLE [db.]name [ON CLUSTER cluster] [PARTITION partition | PARTITION ID 'partition_id'] [FINAL] [DEDUPLICATE [BY expression]]
+    /// ```
+    ///
+    /// See ClickHouse <https://clickhouse.com/docs/en/sql-reference/statements/optimize>
+    OptimizeTable {
+        name: ObjectName,
+        on_cluster: Option<Ident>,
+        partition: Option<Partition>,
+        include_final: bool,
```
Comment (iffyio) [inline src/ast/mod.rs:2843]: can this be called plain `final` (thinking similar to the volatile, temporary etc flags on a CreateTable)?

**Review by iffyio — APPROVED** [review]
Left a couple minor comments re naming but the changes look good to me overall! Thanks!
cc @alamb

File: src/ast/ddl.rs (hunk 3 above)
Comment (git-hulk) [reply src/ast/ddl.rs]: Good call, changed according to the comment.

File: src/ast/mod.rs (hunk 4 above)
Comment (git-hulk) [reply src/ast/mod.rs:2843]: I didn't change `include_final` to `final` since it's a reserved keyword for Rust.

Comment (iffyio) [reply src/ast/mod.rs:2843]: Ah makes sense!

**Review by alamb — APPROVED** [review]
Thank you @git-hulk  and @iffyio  -- way to keep the code flowing

**Pull request merged**